File: Python/examples/detailed_mode_test_app.py
# -*- coding: utf-8 -*-

from __future__ import print_function, unicode_literals

import sys, os, uuid, time
import semantria


# API Key/Secret
# Set the environment vars before calling this program
# or edit this file and put your key and secret here.
consumerKey = os.getenv('SEMANTRIA_KEY')
consumerSecret = os.getenv('SEMANTRIA_SECRET')

# Task statuses
TASK_STATUS_UNDEFINED = 'UNDEFINED'
TASK_STATUS_FAILED = 'FAILED'
TASK_STATUS_QUEUED = 'QUEUED'
TASK_STATUS_PROCESSED = 'PROCESSED'


def onRequest(sender, result):
    print("\n", "REQUEST: ", result)

def onResponse(sender, result):
    print("\n", "RESPONSE: ", result)

def onError(sender, result):
    print("\n", "ERROR: ", result)

def onDocsAutoResponse(sender, result):
    print("\n", "AUTORESPONSE: ", len(result), result)

def onCollectionsAutoResponse(sender, result):
    print("\n", "AUTORESPONSE: ", len(result), result)


print("Semantria Detailed mode demo ...")
print("")

# Initializes new session with the serializer object and the keys.
session = semantria.Session(consumerKey, consumerSecret)

# Initialize session callback handlers
# session.Request += onRequest
# session.Response += onResponse
session.Error += onError
# session.DocsAutoResponse += onDocsAutoResponse
# session.CollectionsAutoResponse += onCollectionsAutoResponse

subscription = session.getSubscription()

initialTexts = []

print("Reading collection from file...")
with open('source.txt', encoding='utf-8') as f:
    for line in f:
        initialTexts.append(line)

results = []
tracker = {}
batch = []
for text in initialTexts:
    # Creates a sample document which need to be processed on Semantria
    # Unique document ID
    # Source text which need to be processed
    doc_id = str(uuid.uuid4())
    batch.append({'id': doc_id, 'text': text})
    tracker[doc_id] = TASK_STATUS_QUEUED
    
    if len(batch) == subscription['basic_settings']['incoming_batch_limit']:
        res = session.queueBatch(batch)
        if res in [200, 202]:
            print("{0} documents queued successfully.".format(len(batch)))
            batch = []

if len(batch):
    res = session.queueBatch(batch)
    if res not in [200, 202]:
        print("Unexpected error!")
        sys.exit(1)
    print("{0} documents queued successfully.".format(len(batch)))

print("")

while len(list(filter(lambda x: x == TASK_STATUS_QUEUED, tracker.values()))):
    time.sleep(0.5)
    print("Retrieving your processed results...")

    response = session.getProcessedDocuments()
    for item in response:
        if item['id'] in tracker:
            tracker[item['id']] = item['status']
            results.append(item)

print("")

# Print sample of analysis results. (There's lots more in there!)
for data in results:
    # Print document sentiment score
    print("Document {0} / Sentiment score: {1}".format(
        data['id'], data['sentiment_score']))

    # Print document themes
    if "themes" in data:
        print("Document themes:")
        for theme in data["themes"]:
            print("\t {0} (sentiment: {1})".format(
                theme['title'].encode('ascii', 'replace'), theme['sentiment_score']))

    # Print document entities
    if "entities" in data:
        print("Entities:")
        for entity in data["entities"]:
            print("\t {0}: {1} (sentiment: {2})".format(
                entity['title'].encode('ascii', 'replace'), entity['entity_type'],
                entity['sentiment_score']
            ))

    print("")

print("Done!")

File: Ruby/examples/detailed_mode_test_app.rb
#!/usr/bin/env ruby
require 'securerandom'
require 'semantria'

print 'Semantria Detailed mode demo ...', "\r\n"

# the consumer key and secret
$consumer_key = ''
$consumer_secret = ''

# Task statuses
$TASK_STATUS_UNDEFINED = 'UNDEFINED'
$TASK_STATUS_FAILED = 'FAILED'
$TASK_STATUS_QUEUED = 'QUEUED'
$TASK_STATUS_PROCESSED = 'PROCESSED'

$initial_texts = []

class SessionCallbackHandler < Semantria::CallbackHandler
  def onRequest(sender, args)
    #print "Request: ", args, "\n"
  end

  def onResponse(sender, args)
    #print "Response: ", args, "\n"
  end

  def onError(sender, args)
    print 'Error: ', args, "\n"
  end

  def onDocsAutoResponse(sender, args)
    #print "DocsAutoResponse: ", args.length, args, "\n"
  end

  def onCollsAutoResponse(sender, args)
    #print "CollsAutoResponse: ", args.length, args, "\n"
  end
end

# Initializes new session with the keys and app name.
# We also will use compression.
session = Semantria::Session.new($consumer_key, $consumer_secret, 'TestApp', true)

# Initialize session callback handlers
callback = SessionCallbackHandler.new()
session.setCallbackHandler(callback)

subscription = session.getSubscription()

documents = []
tracker = Hash.new

print("Reading collection from file...\n")
f = File.open('source.txt').read
f.gsub!(/\r\n?/, "\n")
f.each_line do |line|
  if not line.empty?
    $initial_texts.push(line)
  end
end

if $initial_texts.size < 1
  print("Source file isn't available or blank.\n")
  exit(1)
end

$initial_texts.each do |text|
  # Creates a sample document which need to be processed on Semantria
  # Unique document ID
  # Source text which need to be processed
  doc_id = SecureRandom.uuid

  documents.push({'id' => doc_id, 'text' => text})
  tracker[doc_id] = $TASK_STATUS_QUEUED
  
  if documents.size == subscription['basic_settings']['incoming_batch_limit']
    result = session.queueBatch(documents)
    if result == 200 or result == 202
      print("#{documents.size} documents queued successfully.\n")
      documents = []
    end
  end
end

if documents.size > 0
  result = session.queueBatch(documents)
  if result != 200 and result != 202
    print("Unexpected error!")
    exit(1)
  end
  
  print("#{documents.size} documents queued successfully.\n")
end

print("\n")

results = []

while not tracker.keep_if{|k, v| v == $TASK_STATUS_QUEUED}.empty?
  sleep(0.5)
  print("Retrieving your processed results...\n")

  response = session.getProcessedDocuments
  response.each do |item|
    if tracker.has_key?(item['id'])
      tracker[item['id']] = item['status']
      results.push(item)
    end
  end
end

print("\n")

results.each do |data|
  # Printing of document sentiment score
  print 'Document ', data['id'], ' Sentiment score: ', data['sentiment_score'], "\r\n"

  # Printing of document themes
  print 'Document themes:', "\r\n"
  data['themes'].nil? or data['themes'].each do |theme|
    print "\t", theme['title'], ' (sentiment: ', theme['sentiment_score'], ")", "\r\n"
  end

  # Printing of document entities
  print 'Entities:', "\r\n"
  data['entities'].nil? or data['entities'].each do |entity|
    print "\t", entity['title'], ' : ', entity['entity_type'], ' (sentiment: ', entity['sentiment_score'], ')', "\r\n"
  end

  print "\r\n"
end
